platforms/core-runtime/internal-instrumentation-api/src/main/java/org/gradle/internal/instrumentation/api/jvmbytecode/DefaultBridgeMethodBuilder.java

```diff
@@ -238,5 +302,59 @@ protected void buildBridgeMethodImpl(MethodVisitorScope mv) {
             // The constructor's interceptor returns nothing, so the next <ref> will be used as the return value.
             super.buildBridgeMethodImpl(mv);
         }
+
+        private static String buildConstructorBridgeDesc(String owner, String desc) {
+            // The constructor is represented as NEWINVOKESPECIAL:Owner.<init>(...)V. The bridge method have to call
+            // the constructor itself and return the constructed value. Thus, we change the return value to Owner.
+            Type originalOwner = Type.getObjectType(owner);
+            Type originalMethodType = Type.getMethodType(desc);
+            return Type.getMethodDescriptor(originalOwner, originalMethodType.getArgumentTypes());
+        }
+    }
+
+    private static class InstanceBridgeMethodBuilder extends DefaultBridgeMethodBuilder {
+        private final int tag;
+        private final String originalDesc;
+
+        InstanceBridgeMethodBuilder(int tag, String originalOwner, String originalDesc, String interceptorOwner, String interceptorName, String interceptorDesc) {
+            super(buildInstanceBridgeDesc(tag, originalOwner, originalDesc), interceptorOwner, interceptorName, interceptorDesc);
+            this.tag = tag;
+            this.originalDesc = originalDesc;
+        }
+
+        private InstanceBridgeMethodBuilder(String refinedOwner, InstanceBridgeMethodBuilder builder) {
+            super(builder, buildInstanceBridgeDesc(builder.tag, refinedOwner, builder.originalDesc));
+            this.tag = builder.tag;
+            this.originalDesc = builder.originalDesc;
+        }
+
+        private InstanceBridgeMethodBuilder(InstanceBridgeMethodBuilder builder, boolean hasKotlinDefaultMask, @Nullable String binaryClassName, @Nullable BytecodeInterceptorFilter context) {
+            super(builder, hasKotlinDefaultMask, binaryClassName, context);
+            this.tag = builder.tag;
+            this.originalDesc = builder.originalDesc;
+        }
+
+        @Override
+        protected InstanceBridgeMethodBuilder copy(boolean hasKotlinDefaultMask, @Nullable String binaryClassName, @Nullable BytecodeInterceptorFilter context) {
+            return new InstanceBridgeMethodBuilder(this, hasKotlinDefaultMask, binaryClassName, context);
+        }
+
+        @Override
+        public BridgeMethodBuilder withReceiverType(String targetType) {
+            return new InstanceBridgeMethodBuilder(targetType, this);
+        }
+
+        private static String buildInstanceBridgeDesc(int tag, String owner, String desc) {
+            assert tag == H_INVOKEINTERFACE || tag == H_INVOKEVIRTUAL;
```

Comment:
💅 Should we use `Precoditions` since `assert` are disabled unless run with `-ea` flag?

Reply:
> 💅 Should we use `Precoditions` since `assert` are disabled unless run with `-ea` flag?

This is a private method, so I don't think there is much value in checking that in the field, it is mostly a documentation at this point. I believe we run tests with asserts enabled, so if someone messes up with the switch above, the test suite will catch that.